### DISCUSSION TITLE
libvips/8.15.1 from conan and vips_tiffsave for conversion to pyramidal images

### DISCUSSION THREAD

**Discussion by bkocev** [libvips/libvips#3828] [General]
Dear all,

I am using the C api of libvips.
I was using VIPS 8.5.0 previously and was able to convert large tiff images to pyramidal images using the following code: 

        vips_tiffsave( inputImage,
                       pyramidalTIFFPath,
                       "pyramid",
                       true,
                       "tile",
                       true,
                       "compression",
                       VIPS_FOREIGN_TIFF_COMPRESSION_LZW,
                       "bigtiff",
                       true,
                       NULL )

However, after updating to libvips/8.15.1 from conan, I am getting errors when doing the conversion. Sometimes the whole process fails. Sometimes I get an output where the low resolution levels are messed up and only the highest resolution looks ok. Sometimes, I get this error: 

error buffer: TIFFAppendToStrip: Maximum TIFF file size exceeded
vips2tiff: TIFF write tile failed
wbuffer_write: write failed
TIFFAppendToStrip: Maximum TIFF file size exceeded

Did something change in libvips/8.15.1 from conan? 

Thank you,
Bojan.

**Comment by jcupitt** [comment]
Hello @bkocev,

Oh dear, it should work. 

I tried:

```
$ for i in {1..100}; do echo $i; vips tiffsave wtc.jpg x.tif --pyramid --tile --compression lzw --bigtiff; done
1
2
3
4
...
100
$
```

That's a 10k x 10k JPEG. We don't make the conan package, perhaps there's some problem with their binary? Have you tried any others?

**Comment by bkocev** [comment]
Hello,

For "smaller" (e.g., 13811 X 6122) single resolution images, I managed to fix the problem by switching to version libvips/8.14.2 from conan. So, indeed it is version-related.

For "larger" (e.g., 122880 X 110592) images, the conversion to a pyramidal file fails with this error:

VIPS:ERROR:../src/libvips/iofuncs/target.c:482:vips_target_flush: assertion failed: (target->write_point >= 0)
Bail out! VIPS:ERROR:../src/libvips/iofuncs/target.c:482:vips_target_flush: assertion failed: (target->write_point >= 0)

Is there a limitation on how large the images could be, so that the conversion can work?

Thank you and kind regards,
Bojan.

**Reply by jcupitt** [reply]
I often make 500,000 x 500x000 pixel pyramids, so it should be fine.

I suspect there is an issue with the conan build. Could you try with another libvips binary? Perhaps Debian or Windows?

**Reply by jcupitt** [reply]
I tried:

```
$ vips crop D23001205_26-10-2023_09-53-58.mrxs[rgb] x.tif[tile,pyramid,compression=lzw,bigtiff] 0 0 122880 110592 --vips-progress
vips temp-2: 122880 x 110592 pixels, 32 threads, 128 x 128 tiles, 640 lines in buffer
vips temp-2: done in 212s          
$ vips --version
vips-8.15.1
```

**Reply by bkocev** [reply]
Hi,

I integrated [vips-dev-w64-all-8.15.1.zip](https://github.com/libvips/build-win64-mxe/releases/download/v8.15.1/vips-dev-w64-all-8.15.1.zip) from https://github.com/libvips/build-win64-mxe/releases/tag/v8.15.1.

Unfortunately, I face the same problems as with the conan package. So, I think it has to do with the windows binaries from libvips.

Running this command "vips.exe tiffsave <input image> <output image> --pyramid --tile --compression lzw --bigtiff" result in an output image in which the low resolution levels have messed up colors. The highest resolution level works. Experimenting with different compression types, I could find out that there is a problem with some compressions. This is what I found out:

**Do not work**: deflate, lzw, ccittfax4 (creation fails: Bits/sample must be 1 for Group 3/4 encoding/decoding), zstd (creation fails: ZSTD compression support is not configured)
**Do work**: none, jpeg, packbits
**Not tested:** webp (), jp2k 

Any idea why the lzw does not work?

Thank you and kind regards,
Bojan.

**Reply by bkocev** [reply]
There is a parameter "predictor" in tiffsave, i.e., if I run it as:

vips.exe tiffsave <input image> <output image> --pyramid --tile --compression lzw --bigtiff --predictor none

the issue is gone.

Is it safe/correct to use with "predictor" set to "none"?

**Reply by jcupitt** [reply]
Ahhh! That's a great clue, well done @bkocev.

Yes, this is an issue fixed in 8.15.2, sorry, I should have realized this sooner. 

The fix is here: https://github.com/libvips/libvips/commit/618e2b56c9f9b1f9fe5adc99426ebdcad88de1ce

8.15.2 has some useful stuff now, I'll make a new release.
